### PR TITLE
[spirv] Translate vector types and handle variable initializers

### PR DESCRIPTION
**Pull request by antiagainst**
* Translated vector types and operations on vector types
  * +, -, *, /, % on vector <op> vector and vector <op> scalar
  * Prefix increment on vectors
  * < on vector <op> vector and vector <op> scalar
* Covered both constant initializer and variable initializer
* Applied constant modifier to variables

### CODE REVIEW

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 1]
```diff
@@ -673,12 +747,36 @@ class SPIRVEmitter : public ASTConsumer {
...
+
+      return theBuilder.createCompositeConstruct(vecTypeId, elements);
+    }
+    case CastKind::CK_HLSLVectorToScalarCast: {
+      // We should have already treated vectors of size 1 as scalars.
+      // So do nothing here.
+      if (hlsl::GetHLSLVecSize(subExpr->getType()) == 1) {
+        return doExpr(subExpr);
+      }
+      emitError("vector to scalar cast unimplemented");
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: I'm wondering: Is there any valid case of casting a vector larger than 1 element to a scalar?

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 2]
```diff
@@ -802,6 +949,17 @@ case BO_##kind : {                                                             \
       return theBuilder.getConstantFloat32(1.0);
     }
 
+    if (hlsl::IsHLSLVecType(type)) {
+      const QualType elemType = hlsl::GetHLSLVecElementType(type);
+      const uint32_t elemOneId = getValueOne(elemType);
+
+      const size_t size = hlsl::GetHLSLVecSize(type);
+      llvm::SmallVector<uint32_t, 4> elements(size, elemOneId);
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: don't you need to handle the `vec1` corner case here? ConstantComposite shouldn't be used for vector of size 1

**Comment by ehsannas** [comment]
The rest looks good to me! 👍

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 2 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Good catch! Actually this was fixed in a later CL: https://github.com/antiagainst/DirectXShaderCompiler/commit/58aef4471ca614a888df767b22893a6095591f4e#diff-d00a9d7516b840ab223da9e8420a9844R1181

I'll move that change forward to this CL.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 1 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Yes. That's actually a truncation. When you assign a vector to a scalar, only the first element got preserved and written into the scalar.

Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: See https://github.com/antiagainst/DirectXShaderCompiler/commit/6caf1a9229555433bc171ea173cc691ac98220dc

Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Interesting! +1

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 2 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Done

**Pull request merged**